alphabase/spectral_library/annotate.py

```diff
@@ -0,0 +1,557 @@
+        "peak_start_idx"
+    ].astype(int)
+    merged_precursor_df["peak_stop_idx"] = merged_precursor_df["peak_stop_idx"].astype(
+        int
+    )
+
+    # Create a new SpecLibFlat object with the updated precursor DataFrame
+    annotated_speclib_flat = speclib_flat.copy()
+    annotated_speclib_flat.precursor_df = merged_precursor_df
+
+    return annotated_speclib_flat
+
+
+def annotate_fragments_flat(
+    speclib_flat: SpecLibFlat,
+    raw_data_peak_df: pd.DataFrame,
+    mass_error_ppm: float = 20,
+) -> SpecLibFlat:
+    """Annotate a spectral library flat with raw data and calculate PIF.
+
+    This function matches observed spectrum peaks to theoretical fragments,
+    calculates the precursor ion fraction (PIF), and organizes the results
+    into DataFrames.
+
+    Parameters
+    ----------
+    speclib_flat : SpecLibFlat
+        A spectral library flat object containing precursor and fragment information.
+    raw_data_peak_df : pd.DataFrame
+        Raw data object containing peak information.
+    mass_error_ppm : float, optional
+        Mass error tolerance in parts per million (default: 20.0).
+
+    Returns
+    -------
+    SpecLibFlat
+        An updated SpecLibFlat object with annotated fragment information.
+
+    """
+    if not all(
+        col in speclib_flat.precursor_df.columns
+        for col in ["peak_start_idx", "peak_stop_idx"]
+    ):
+        raise ValueError(
+            "peak_start_idx and peak_stop_idx must be present in speclib_flat.precursor_df. Please run annotate_precursors_flat() first."
+        )
+
+    fragment_df_list = []
+    matched_precursor_df = speclib_flat.precursor_df.copy()
+    matched_precursor_df["pif"] = 0.0
+
+    start_index = 0
+    for i, (peak_start, peak_stop, flat_frag_start, flat_frag_stop) in enumerate(
+        zip(
+            matched_precursor_df["peak_start_idx"],
+            matched_precursor_df["peak_stop_idx"],
+            matched_precursor_df["flat_frag_start_idx"],
+            matched_precursor_df["flat_frag_stop_idx"],
+        )
+    ):
+        spectrum_peak_df = raw_data_peak_df.iloc[peak_start:peak_stop].copy()
+        fragment_df = speclib_flat.fragment_df.iloc[flat_frag_start:flat_frag_stop]
+
+        (
+            spectrum_peak_df["type"],
+            spectrum_peak_df["loss_type"],
+            spectrum_peak_df["charge"],
+            spectrum_peak_df["number"],
+            spectrum_peak_df["position"],
+            spectrum_peak_df["error"],
+        ) = annotate_spectrum(
+            spectrum_peak_df["mz"].to_numpy(),
+            fragment_df["mz"].to_numpy(),
+            fragment_df["type"].to_numpy(),
+            fragment_df["loss_type"].to_numpy(),
+            fragment_df["charge"].to_numpy(),
+            fragment_df["number"].to_numpy(),
+            fragment_df["position"].to_numpy(),
+            mass_error_ppm=mass_error_ppm,
+        )
+
+        matched_precursor_df.loc[i, "pif"] = calculate_pif(spectrum_peak_df)
```

Comment:
Sorry, that I'm asking you to fix my mistakes here :D

could you change the `calulate_pif()` function so it's structured and applied in the same way as the other metrics?
So it's called after annotation and operates on the precursor df.